### PR TITLE
feat(vara,gear-bank): Split gas and fees to author/treasury

### PR DESCRIPTION
**Pull request by ukint-vs**
- Implement splitting of the tx fees to the block author and to the Treasury.
- Add splitting gas fees in gear-bank.
- Add `SplitGas` and `SplitTxFeeRatio` in gear-bank config
@gear-tech/dev

### CODE REVIEW

File: runtime/common/src/lib.rs [hunk 1]
```diff
@@ -79,23 +79,40 @@ pub type NegativeImbalance<T> = <pallet_balances::Pallet<T> as Currency<
...
     <R as frame_system::Config>::AccountId: From<AccountId>,
     <R as frame_system::Config>::AccountId: Into<AccountId>,
 {
     fn on_unbalanceds<B>(mut fees_then_tips: impl Iterator<Item = NegativeImbalance<R>>) {
         if let Some(fees) = fees_then_tips.next() {
-            if let Some(author) = <pallet_authorship::Pallet<R>>::author() {
-                <pallet_balances::Pallet<R>>::resolve_creating(&author, fees);
-            }
+            // for fees, 50% to treasury, 50% to author
+            let mut split = fees.ration(50, 50);
```
Comment (gshep) [inline runtime/common/src/lib.rs:108]: i would rename to something like `let (mut to_author, to_treasury) = ...`

File: runtime/common/src/lib.rs [hunk 2]
```diff
@@ -79,23 +79,40 @@ pub type NegativeImbalance<T> = <pallet_balances::Pallet<T> as Currency<
...
             if let Some(tips) = fees_then_tips.next() {
-                if let Some(author) = <pallet_authorship::Pallet<R>>::author() {
-                    <pallet_balances::Pallet<R>>::resolve_creating(&author, tips);
-                }
+                // for tips, if any, 100% to author
+                tips.merge_into(&mut split.1);
             }
+            use pallet_treasury::Pallet as Treasury;
```
Comment (gshep) [inline runtime/common/src/lib.rs:113]: please place the import in the function start

**Review by gshep — APPROVED** [review]

**Review by breathx — CHANGES_REQUESTED** [review]
If I'm not mistaken, we were talking about gas fees, not tx-s (or not only tx-s).
And I'd make this constant customisable with governance not only on upgrades.

Furthermore, this question is still considering, so pls don't merge yet.

**Comment by breathx** [comment]
Is it ready for review?

**Comment by ukint-vs** [comment]
> Is it ready for review?

Not yet. Need final touches and tests

**Comment by breathx** [comment]
Fix tests to pass CI pls @ukint-vs

**Comment by ukint-vs** [comment]
@CodiumAI-Agent /review

**Comment by breathx** [comment]
@ukint-vs still fails CI

**Comment by ukint-vs** [comment]
> @ukint-vs still fails CI

Fixed. Set 100% to the block author

**Comment by breathx** [comment]
To be reviewed soon.

Please consider to impl 3 different constants in code in some dedicated place, for example - runtime/common.
Percents from tx fees; percents from gas burned; percents from storage rent gas.

**Comment by breathx** [comment]
According to @ukint-vs, WIP

**Comment by breathx** [comment]
@ukint-vs pls finalise this one

File: runtime/vara/src/integration_tests.rs [hunk 3]
```diff
@@ -618,3 +618,48 @@ fn slashed_proposals_back_to_treasury() {
...
+            let fee = Balances::issue(10);
+            let tip = Balances::issue(20);
+
+            assert_eq!(
+                Balances::free_balance(Treasury::account_id()),
+                EXISTENTIAL_DEPOSIT
+            );
+            assert_eq!(Balances::free_balance(alice.to_account_id()), STASH);
+
+            DealWithFees::on_unbalanceds(vec![fee, tip].into_iter());
```
Comment (breathx) [inline runtime/vara/src/integration_tests.rs:655]: shouldn't the case be covered by pallet transaction payment?

File: runtime/vara/src/lib.rs [hunk 4]
```diff
@@ -1021,12 +1021,16 @@ parameter_types! {
     pub Schedule: pallet_gear::Schedule<Runtime> = Default::default();
     pub BankAddress: AccountId = BANK_ADDRESS.into();
     pub const GasMultiplier: common::GasMultiplier<Balance, u64> = common::GasMultiplier::ValuePerGas(VALUE_PER_GAS);
+    pub SplitGas: Option<(Perbill, AccountId)> = None;
```
Comment (breathx) [inline runtime/vara/src/lib.rs:1024]: please open todo to move this logic into pallet storage to be able to adjust the value by governance avoiding runtime upgrades

**Review by breathx — CHANGES_REQUESTED** [review]
On gear bank side, please perform ratio multiplying only once before transfer
Otherwise, great done 👍🏻

**Comment by breathx** [comment]
Actualise branch and resolve opened discussions, please @ukint-vs

File: pallets/gear-bank/src/lib.rs [hunk 5]
```diff
@@ -46,6 +46,8 @@ macro_rules! impl_config {
             type Currency = Balances;
             type BankAddress = BankAddress;
             type GasMultiplier = GasMultiplier;
+            type SplitGas = SplitGas;
```
Comment (breathx) [inline pallets/gear-bank/src/lib.rs:49]:
```suggestion
            type SplitGasFeeRatio = SplitGasFeeRatio;
```

File: runtime/vara/src/integration_tests.rs [hunk 6]
```diff
@@ -861,3 +861,48 @@ fn fungible_api_works() {
             );
         });
 }
+
+#[test]
+fn test_fees_and_tip_split() {
```
Comment (breathx) [inline runtime/vara/src/integration_tests.rs:866]: Test for gas would be awesome also

**Review by breathx — APPROVED** [review]

File: runtime/vara/src/integration_tests.rs (hunk 6 above)
Comment (ukint-vs) [reply runtime/vara/src/integration_tests.rs:866]: test case is in gear-bank tests

**Pull request merged**